Fixes: Menu on canvas, open in new tab and external image in LivePreview

Mara-Li opened the pull request:
The duplicate menu on canvas bored me a lot, so I removed it from the canvas, and added the option for copy directly in the context menu on canvas card.

For external image, I do the same but I use the node-menu.

I only added support for copy as other option are already supported by Obsidian & canvas.

Moreover, I made also a temporary fix for #35 with adding the copy option in the url-menu.
Also fix #39 

Pretty sorry for the whitespace and some other things. As you didn't set anything about formatting in eslint, my default settings overriding your writing.

Mara-Li committed:
fix: menu on canvas broken

Mara-Li committed:
fix: external image copy in clipboard (live preview)

Mara-Li committed:
fix: use default open tab for image

Mara-Li committed:
refactor: remove unused

Mara-Li committed:
refactor: remove comments

NomarCub commented:
Hi!

Thank you very much for the fixes!

> The duplicate menu on canvas bored me a lot

Could you describe this issue in detail? I don't have a ticket for this yet, and I don't really use the canvas. With screenshots, before and after, preferably. I'd like to know what the issue was, and how the new code fixes it. And if possible for all the fixes you did (if not otherwise obvious in the links). Sorry for the nitpicking, but I'd like to understand the code properly for future maintenance.

> For external image, I do the same but I use the node-menu.

What do you mean by node menu? By "do the same", do you mean only on the canvas?

As for the formatting: I don't mind it in general, but it makes the changes really hard to see in this case. Could you at least rebase, split or reorder the commits somehow, so that some commits are formatting only, and the others contain no formatting changes?
And how did you do the formatting? If it was through something like ESLint or Prettier, or some other CLI then I can reproduce it more easily on my end. I just used VSCode's defaults. It could also help in doing a formatting commit first on my end, so the changes are easier to see.

Mara-Li commented:
First, for the part on canvas : The menu always open on the context menu, so it causes issues when manipulating image (and I work a lot with image in canvas).

Node-Menu is a function in canvas that seems to "trigger" on a right click on canvas, as context menu for canvas doesn't works for external links.

Here some preview in my own version:
![image](https://github.com/NomarCub/obsidian-copy-url-in-preview/assets/30244939/e7d2a61b-a5d6-4a17-8108-f8457c262f22)
VS your:
![image](https://github.com/NomarCub/obsidian-copy-url-in-preview/assets/30244939/1b31ee85-bdc2-4c3d-80fd-6648d6b7c761)

(+ some option are redondant, as they are already handled in canvas)

---
About the commit, I'm sorry, I commited "too quickly" and I don't think I can do something, unless rebase?

It could be useful to have some eslint with a formatting, and I could take it from you and apply. So the problem "won't be too much".
Also, in the diff, you can hide whitespace change here: 
![image](https://github.com/NomarCub/obsidian-copy-url-in-preview/assets/30244939/5f45fc78-86c5-44be-95af-22eaaee109d9)

I will try to re-use default vscode settings in hoping it will fixes some parsing!

Mara-Li commented:
Oh also, what tab size (or space?) do you use? I think it will fix a lot!

Mara-Li committed:
lint: formatting issue

Mara-Li committed:
lint: adjust formating to be more accurate with main repo

NomarCub commented:
Thanks for the formatting changes! It'll be easier for me now.

What are the repro steps for the double canvas menu? I tried a canvas with directly embedded images, notes with images - both reading and live preview, cards, but no right click yielded the double menu on your screenshot.

NomarCub commented on src/helpers.ts:
```diff
@@ -1,114 +1,106 @@
-    app.workspace.setActiveLeaf(leaf, { focus: true });
+	const activeFile = app.workspace.getActiveFile();
+	const imageAsTFile = activeFile
+		? app.metadataCache.getFirstLinkpathDest(image.alt, activeFile.path)
```
Why did you use `image.alt` here?
According to web docs, this is not what it's for: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt.
I guess it might work here, if Obsidian sets the path as alt text, but the user can override that for an image in inline HTML or even regular markdown, which would break the functionality.
Was the `getRelativePath` method in this helper file not fit for this purpose? I know it's not robust either, and I'd like to make it better, and I'd like to know your opinion on this.

Mara-Li replied:
You're right! I totally forgot alias 🤦 and check the methods you write !
Fixed :>

NomarCub committed:
slight formatting tweaks

NomarCub committed:
remove onunload method

NomarCub committed:
extract some common strings to a `strings` object

NomarCub commented on src/main.ts:
```diff
@@ -15,327 +15,406 @@ const deleteTempFileTimeout = 60_000;
+						.setTitle(strings.menuItems.copyImageToClipboard)
+						.onClick(async () => {
+							const imageBuffer = await this.app.vault.readBinary(file);
+							const blob = new Blob([imageBuffer], { type: "image/png", });
```
Please de-duplicate this block and others starting with `const blob =`.

Mara-Li replied:
Done

Mara-Li commented:
Pretty strange you can repro my issue. I have it from month on multi vault!

I just set a image in the canvas, like adding image. Not image in a note.

NomarCub commented:
Ok, I found it. The double menu only happens, when the image is already selected.
I also made two comments on the code I'd like you to look at.

Mara-Li commented:
Yeah, sure, where?

NomarCub commented on src/main.ts:
```diff
@@ -15,327 +15,406 @@ const deleteTempFileTimeout = 60_000;
+		);
+		this.registerEvent(
+			this.app.workspace.on("canvas:node-menu", (menu, node) => {
+				//@ts-ignore
```
Please make an interface for this, like the ones in helpers.ts, instead of ts-ignore.

Mara-Li committed:
fix: don't use image.alt as it is overwriten by alias

Mara-Li committed:
fix: deduplicate block for blob

Mara-Li committed:
fix: extends canvas in helper.ts

Mara-Li commented:
BTW, it could be cool to insert some translation support, but i don't know if you want to use i18n or no library for that, as they are not so much string to translate.

NomarCub committed:
update obsidian-typings, npm audit fix

NomarCub committed:
make esbuild ignore commented (// or #) lines in .devtarget

NomarCub committed:
remove disabled lint rules with no effect

NomarCub committed:
reformat file, restructure switch in onImageContextMenu

NomarCub committed:
update dependencies

NomarCub committed:
optionally re-enable context menu on canvas

NomarCub committed:
bump version to 1.7.3

NomarCub committed:
rename function to copyImageToClipboard

NomarCub committed:
reformat

NomarCub commented:
Thank you for the changes, nice solutions!

I've made some modifications. Everything should still work, I made mostly minor changes to your part of the code. Would you please take a last look at the code and the functionality? I tried to separate commits so the changes are easier to see.
If you have any advice on other stuff for this plugin, do let me know, this project is kind of a mess 😄

I also added an option to re-enable the duplicate menu. It's faulty, but it might come handy for images in notes on canvas still.

As for translation: I'm down to give it a shot, if there's a demand for it and people make contributions for new languages.

NomarCub committed:
format eslintrc

Mara-Li committed:
fix: allow context menu for image embedded in card without duplicate menu on loaded image

Mara-Li commented:
For the new option in settings, it's a good idea, because I totally forgot about image embedded.
But, as I worked (too much) on canvas, I know a way to have a good discrimination between embedded image in a card and loaded image : the parent have a class to discriminate!
With this, I don't know if we need to continue to have this option in settings? I left it so you can remove it or keep it as you want :)

Mara-Li commented:
I also noticed a bug for PDF when the option is enabled, so I will try to fix it

Mara-Li committed:
fix(canvas): prevent crash on loaded embed

Mara-Li commented:
On canvas, it seems that the menu is not triggered properly for embedded pdf 🤔 
Seems to be because of this: 
https://github.com/NomarCub/obsidian-copy-url-in-preview/blob/119496d4c5523ab3e330d9c58a882f2fd353f36b/src/main.ts#L131-L137

Mara-Li committed:
fix(pdf): on canvas, hide the default menu (prevent duplication) and reopen it when the pdf-menu disappear.

Mara-Li committed:
refactor: rename variable for otherMenu

Mara-Li commented:
I hope this is good for you :>

NomarCub committed:
formatting

NomarCub committed:
restructure code among files

NomarCub committed:
add `Open in new tab` to canvas menu

NomarCub committed:
introduce setMenuVisuals, move registerEscapeButton to helpers

NomarCub committed:
move clipboard to setMenuItem

NomarCub commented:
Thanks for the changes, again!
I made some further restructuring, and added an Open in new tab menu to canvas.
Would you please take another look at the functionality? I hope I didn't ruin anything. Also let me know if you have any more recommendations.

NomarCub committed:
simplify withTimeout - see: https://blog.logrocket.com/when-to-use-never-unknown-typescript/

Mara-Li commented:
For PDF, why we need a menu tbh ?
![image](https://github.com/NomarCub/obsidian-copy-url-in-preview/assets/30244939/df74226c-3e3e-407d-98f1-899997d34e64)
Now, pdf can be directly opened with this menu, even in LivePreview?

("Ouvrir avec l'application par défaut = open with default application)

Oh, also, I discover how to use the default Obsidian translation while working on another plugin, so I will implement it for the menu, as we use the same word (and commands)

Mara-Li committed:
feat: translation support
Use the default translation of Obsidian for a max fun!

Mara-Li committed:
Some fix around canvas
- prevent opening duplicate menu in image (loaded)
- Add command for loaded pdf too

Mara-Li committed:
feat: add sections in image menu

Mara-Li committed:
refactor: set section in setMenuItem directly

Mara-Li committed:
refactor: use obsidian default icon for context menu pdf

Mara-Li commented:
To be honest, I'm so happy for this works! I discover new things, and added a lot of things that seems to be default Obsidian !
Also, you are a great developper, it's a pleasure to works with you :>

NomarCub committed:
npm i

NomarCub committed:
reformat

NomarCub committed:
reformat

NomarCub commented:
The PDF menu was added by a contributor before the Obsidian native support.
I never really used it and don't know if people need it anymore, I'll probably remove support eventually.

Nice work on the translation, I didn't know this! I'll check the functionality sometime later.

Why were these two needed for the translation?
- https://github.com/NomarCub/obsidian-copy-url-in-preview/pull/40/files#diff-23c7d6d050a365564a07eeef435d1d9fc8ea94230a751483afe643435cc9647dR23
- https://github.com/NomarCub/obsidian-copy-url-in-preview/pull/40/files#diff-ca5696b367d474e785317cb7a0d9853fef5729387ab8d0fab4c46268c03aae99R132

Mara-Li commented:
Tbh, I don't know why but I know we need it, to be recognized using Obsidian :'D
At last we can try without it. It's just, I always declare i18n in types and it's in the docs for i18next. We need to use this particular library because Obsidian use it. 
https://www.i18next.com/overview/typescript

NomarCub commented:
So the global is from Obsidian, right?
Was the bind and cast to unknown the only option to call it? I looked at `t`'s declarations and I'm not sure how to call it properly, but the current way is kinda messy.
Also what is this other plugin where you discovered Obsidian translations? I'm curious.

Mara-Li commented:
Sorry for the late reply (I was home the all day xD)
It was bartender, a plugin not in community store.

I don't think it's more messy than hacking the click to add a menu XD

But, I hope, in future, that the API allow to translate directly, and get the translation value without using `bind` from i18next.

NomarCub committed:
use i18next directly

NomarCub committed:
bump version to 1.8.0 instead

NomarCub committed:
add Mara-Li to credits

NomarCub committed:
rename to setMenuItem

NomarCub committed:
rename to setMenuItem

NomarCub commented:
I managed to use `i18next.t` directly, I hope that's enough.
Everything seems to work, I added you to the credits too 😄
If it's fine by you, I'll make a release soon.

Mara-Li commented:
Incredible it works out of the box like that!
I learned somethings :>

You can merge, everything works !

NomarCub committed:
simplify translation

NomarCub committed:
update readme

NomarCub commented:
I just released 1.8.0 after merging this. Thank you for being so ready to collaborate and change things over and over, and for all the new things you brought to this plugin! It was a pleasure working with you 😌

NomarCub commented:
Hey @Mara-Li!
I'm not sure what is the best way to reach you, this was my first guess.
I made some refactors, and have one question about your code. Would you please group these logical operations with parentheses? I have an ESLint warning about them.

https://github.com/NomarCub/obsidian-copy-url-in-preview/blob/5d85386dc61680b248e7639543f270d2a8513556/src/main.ts#L274-L278

Operator precedence would mean it's supposed to be this:
```ts
if ((!this.settings.enableDefaultOnCanvas && this.app.workspace.getActiveFile()?.extension === "canvas")
	|| event.targetNode?.parentElement?.className === "canvas-node-content media-embed image-embed is-loaded") {
	return;
}
```

But I wanted to be sure about the original intention, I didn't understand this fully.

Mara-Li commented:
Hello!
Sorry for the late reply, I got a new job and my energy was so low :')

The goal of this specific condition is to disable the context menu if the image is "loaded" inside a canvas, meaning drag'n'drop direct image (not embeded in a card).

So, what I wrote is:
- Verify if the added menu (`enableDefaultOnCanvas`) is enabled/disabled
- Verify if we are on a canvas (or not)
- Verify the class of the parentElement to check if it's a image-embed is-loaded.

If none of the condition match, like : 
- The menu is disabled 
- AND we are on a canvas
- OR we are on a loaded image
We return because we don't need to enable lthe menu.

Embeded image directly looks like that in the DOM : ![image](https://github.com/NomarCub/obsidian-copy-url-in-preview/assets/30244939/1ef9abd7-7e5d-4c56-a046-2eba672e535c)

Versus an image in a card: 
![image](https://github.com/NomarCub/obsidian-copy-url-in-preview/assets/30244939/60b900c5-bc2a-4a99-a447-02166049731b)

So you can put parenthesis without worrying!

NomarCub commented:
Thanks!
